src/gululog_idx.erl

```diff
@@ -234,6 +235,34 @@ delete_from_cache(Tid, LogId) ->
   end,
   Tid.
 
+%% @doc Truncate after given logid from cache and index file
+%% Return new index and index file for deleted or truncated
+%% @End
+-spec truncate(dirname(), index(), segid(), logid(), ?undef | dirname()) ->
+  {index(), [filename()]}.
+truncate(Dir, #idx{tid = Tid} = Idx, SegId, LogId, BackupDir) ->
+  Ms = ets:fun2ms(fun(?ETS_ENTRY(_, LogIdX, _) = EtsEntry) when LogIdX > LogId -> EtsEntry end),
+  case EtsEntryList = ets:select(Tid, Ms) of
+    [] ->
+      {Idx, []};
+    _ ->
+      [ets:delete(Tid, LogIdX) || {LogIdX, _} <- EtsEntryList],
+      {TruncateList, DeleteList} =
+        lists:partition(fun(?ETS_ENTRY(SegIdX, _, _)) -> SegIdX == SegId end, EtsEntryList),
+      %% close writer fd
+      flush_close(Idx),
+      %% delete idx file for > segid
+      DeleteResult = truncate_delete_do(DeleteList, Dir, BackupDir),
+      %% truncate idx file for = segid
+      TruncateResult =
+        case TruncateList of
+          []-> [];
+          _ -> truncate_truncate_do(Dir, SegId, LogId, BackupDir)
+      end,
+      %% re-open writer fd
+      {init(Dir), DeleteResult ++ TruncateResult}
```

Comment:
if you do it like this,  there is a new ets table created (the old one is leaking).
also, init/1 is expensive, it's only to be called in start/restart.
here, it should be enough to just open a new writer fd on the latest .idx file.